File: src/remote/sheetsAdapter.ts
import type {
  Event,
  NotificationEntry,
  Parent,
  RideAssignment,
} from "@/domain/models";
import type { GroupSharedConfig } from "@/domain/config";

/**
 * Abstract Sheets adapter. Each group has a shared sheet with tabs for
 * Events, Assignments, Parents, GroupConfig, Notifications.
 */
export interface SheetsAdapter {
  ensureSheetExists(groupId: string, groupName: string): Promise<string>;

  readEvents(sheetId: string): Promise<Event[]>;
  writeEvents(sheetId: string, events: Event[]): Promise<void>;

  readAssignments(sheetId: string): Promise<RideAssignment[]>;
  writeAssignments(sheetId: string, assignments: RideAssignment[]): Promise<void>;

  readParents(sheetId: string): Promise<Parent[]>;
  writeParents(sheetId: string, parents: Parent[]): Promise<void>;

  readGroupConfig(sheetId: string): Promise<GroupSharedConfig>;
  writeGroupConfig(sheetId: string, config: GroupSharedConfig): Promise<void>;

  readNotifications(sheetId: string): Promise<NotificationEntry[]>;
  appendNotification(sheetId: string, entry: NotificationEntry): Promise<void>;
}

File: app/src/main/java/i/drive/kids/data/remote/SheetsAdapter.kt
package i.drive.kids.data.remote

import i.drive.kids.config.GroupSharedConfig
import i.drive.kids.domain.model.Event
import i.drive.kids.domain.model.GroupMember
import i.drive.kids.domain.model.NotificationEntry
import i.drive.kids.domain.model.RideAssignment

interface SheetsAdapter {
    /** Read all Events from the Events tab. */
    suspend fun readEvents(sheetId: String): List<Event>

    /** Read all RideAssignments from the Assignments tab. */
    suspend fun readAssignments(sheetId: String): List<RideAssignment>

    /** Read all parent members from the Parents tab. */
    suspend fun readParents(sheetId: String): List<GroupMember>

    /** Read group-level configuration from the GroupConfig tab. */
    suspend fun readGroupConfig(sheetId: String): GroupSharedConfig

    /** Read all NotificationEntries from the Notifications tab. */
    suspend fun readNotifications(sheetId: String): List<NotificationEntry>

    /** Overwrite all rows in the Events tab. */
    suspend fun writeEvents(sheetId: String, events: List<Event>)

    /** Overwrite all rows in the Assignments tab. */
    suspend fun writeAssignments(sheetId: String, assignments: List<RideAssignment>)

    /** Overwrite the GroupConfig tab with the given config. */
    suspend fun writeGroupConfig(sheetId: String, config: GroupSharedConfig)

    /** Append a single notification entry to the Notifications tab. */
    suspend fun appendNotification(sheetId: String, entry: NotificationEntry)

    /**
     * Ensures a shared Sheet exists for the group.
     * Creates a new spreadsheet with the required tabs if absent.
     * @return the Google Sheets spreadsheetId.
     */
    suspend fun ensureSheetExists(groupId: String, groupName: String): String
}
